Re-factor code to use a Dict-based interface

WolfByttner opened the pull request.

WolfByttner committed:
Add partial interface conversion

WolfByttner committed:
Fix remaining tests

WolfByttner committed:
Fix data packing issue in init function

WolfByttner committed:
Update TrajectorySimulator code

WolfByttner committed:
Merge api-work into integration branch

WolfByttner committed:
Merge pull request #39 from patrickfletcher/api-work-implementation.merge

Merge api-work into implementation branch

patrickfletcher commented:
I keep getting the following type of error when trying to use a callable as input to simulators:
```
TypeError: initialize(): incompatible function arguments. The following argument types are supported:
    1. (self: clode.cpp.clode_cpp_wrapper.SimulatorBase, tspan: List[float], x0: List[float], pars: List[float], solver_params: clode.cpp.clode_cpp_wrapper.SolverParams) -> None

Invoked with: <clode.cpp.clode_cpp_wrapper.SimulatorBase object at 0x000002DC3D7012F0>, (0.0, 10.0), array([dict_values([1.0, 1.0, 1.0])], dtype=object), array([dict_values([27.0, 10.0, 2.6666666666666665])], dtype=object), <solver_params(dt=0.010000, dtmax=1.000000, abstol=0.000001, reltol=0.001000, max_steps=1000000, max_store=1000000, nout=1)>
```

It seems the default values are passed as dict_values, not as list?  

I strongly prefer simplifying the initializer to force only default parameter names/values (no ranges).  I think setting up the ranges is a separate concept - the user needs to precisely specify what they want to do, and should set this the ensemble of parameter + initial_state as matrices using the "set_problem_data" method. 

Providing convenience options hidden in the constructor of the Simulator classes is just going to confuse things.

Let me know if I'm missing something here...

WolfByttner committed:
Add ensemble initialisers

WolfByttner committed:
Fix data packing at return step

WolfByttner committed:
Add OpenCL builtins

WolfByttner committed:
Merge branch 'api-work' into api-work-implementation

patrickfletcher committed:
simulator minor tweaks

set_ensemble: use self._integrator.set_problem_data(vars_array, pars_array)

_cl_array_length --> _ensemble_size

make feature/trajectory interface consistent with base solver (default var/par dicts only)

patrickfletcher committed:
add examples corresponding to docs

patrickfletcher commented:
I modified a few small things:
- the features/trajectory api should expect only default parameter/variable dicts, as in the base solver
- Once initialized by the constructor, we don't need to initialize again in set_ensemble. If the integrator is initialized, we can use set_problem_data.
- I changed the name _cl_array_length --> _ensemble_size.  The true array length in the c++ layer is ```_ensemble_size * nVar``` or ```_ensemble_size * nPar```

This is nearly there - I'm having trouble getting meaningful outputs from the new "spike_counting.py" example. The output results don't make sense, and the feature solver reports that it took zero steps. The other examples seem to work fine... Maybe it's how I'm setting up the data passed to set_ensemble? Not sure what the source of the regression is, but I need to shift gears now to other work.

WolfByttner committed:
Start implementing OpenCL builtin support

WolfByttner committed:
Add OpenCL builtins

WolfByttner committed:
Use List[] rather than list[] in tests for 3.8 compatibility

WolfByttner committed:
Turn float into realtype when casting

WolfByttner committed:
Merge pull request #40 from patrickfletcher/api-work-implementation.builtins

Api work implementation.builtins

WolfByttner committed:
Make set_ensemble not reorder data

patrickfletcher committed:
fix trajectory off-by-one

WolfByttner committed:
Add set_ensemble reorder tests

WolfByttner committed:
Fix n_stored index for aux variables

patrickfletcher committed:
fix nAux=0: needs a dummy auxName

patrickfletcher committed:
some observer cleanup

patrickfletcher committed:
remove spdlog newlines in clODE.cpp

patrickfletcher committed:
add spike-counting example to docs

patrickfletcher committed:
smaller grid for spike_counting.md

patrickfletcher committed:
Create test_observers.py

patrickfletcher committed:
update featureSimulator parameter defaults

patrickfletcher committed:
log "using observer" to debug instead of info

patrickfletcher committed:
clean up & debug observers

patrickfletcher committed:
Better fix for nAux=0, nWiener=0

patrickfletcher committed:
pytest skip test_observers.py
